File: projects/06/quadraticRoots.rb
# quadraticRoots.rb computes the roots of a quadratic equation.
#
# Input: a, b, c, the three factors that are ints.
# Output: root1, root2, the two roots that are ints.
#
# Student Name: ZeAi Sun
# Date: Feb 19, 2023
#################################################################

# quadraticRoots() computes the roots of a quadratic equation.
# Receive: a, b, c, the three factors.
# Return: root1, root2, the two roots.
######################################################

def quadraticRoots(a, b, c)
    arg = b * b - 4 * a * c             # compute the value of b^2-4ac.
    if (a != 0) and (arg >= 0)          # check if a is 0 or b^2-4ac is greater than 0.
        root1 = (-1 * b + Math.sqrt(arg)) / (2 * a);        # if so, compute the roots.
        root2 = (-1 * b - Math.sqrt(arg)) / (2 * a);
    else
        root1 = 0                                           # if not, set the roots to be 0.
        root2 = 0
    end
    result = [root1, root2]                                 # return a list contained the roots.
end

def main
  print "Enter the value of a: "; 
  valueA = gets.to_i
  print "Enter the value of b: "; 
  valueB = gets.to_i
  print "Enter the value of c: "; 
  valueC = gets.to_i
  result = quadraticRoots(valueA, valueB, valueC)
  puts "The first root is: " << result[0].to_s              # print the first root.
  puts "and the second root is: " << result[1].to_s         # print the second root.
end

main

File: projects/06/clojure/src/quadraticRoots.clj
;;;; quadraticRoots.clj computes the roots of a quadratic equation.
;;;;
;;;; Input: a, b, c, the three factors that are ints.
;;;; Output: root1, root2, the two roots that are ints.
;;;;
;;;; Usage: clojure -m quadraticRoots
;;;;
;;;; Student Name: ZeAi Sun
;;;; Date: Feb 20, 2023
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns quadraticRoots)      ; name the program

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; quadraticRoots() computes the roots of a quadratic equation.
;;; Receive: a, b, c, the three factors.
;;; Return: root1, root2, the two roots.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn quadraticRoots [a b c]
    (let [arg (- (Math/pow b 2) (* 4 a c))]
        (if (and (not= a 0) (>= arg 0))     ; check if a is not 0 and b^2-4ac is greater than 0.
            (let [root1 (/ (+ (- 0 b) (Math/sqrt arg)) (* 2 a)) root2 (/ (- (- 0 b) (Math/sqrt arg)) (* 2 a))]
                 (vector root1 root2)       ; if so, compute the roots and put them into a vector to return.
            )
            (vector 0 0)                    ; if not, return 0 for both roots.
        )
    )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; main function
;;; Input: a, b, c, the three factors.
;;; Output: root1, root2, the two roots.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn -main []
  (print "Enter the value of a: ") (flush)
  (let
    [ A (read) ]                      ; read the value of a
    (print "Enter the value of b: ") (flush)
    (let 
        [ B (read) ]                  ; read the value of b
        (print "Enter the value of c: ") (flush)
        (let
            [ C (read)                ; read the value of c
              result (quadraticRoots A B C)
              root1 (get result 0)
              root2 (get result 1)
            ]

            (printf (str "\nThe first root is \"%s\" and " 
                    "the second root is \"%s\"\n\n")
                    root1 root2)
        )
    )
  )
)

